### PR TITLE
feat(react-utilities): creates useSelection hook

### PR DESCRIPTION
**Pull request by bsunderhus**
## New Behavior

1. moves selection manager implementation from `@fluentui/react-table` to `@fluentui/react-utilities` to be reused
2. updates `@fluentui/react-table` accordingly

### CODE REVIEW

**Comment by fabricteam** [comment]
## [Perf Analysis (`@fluentui/react-components`)](https://github.com/microsoft/fluentui/wiki/Perf-Testing)

  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount_base.html">9</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount.html">11</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr></table><details><summary>All results</summary><p>
  <table>
  <tr>
    <th>Scenario</th>
    <th>Render type</th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">Master Ticks</a>
    </th>
    <th>
      <a href="https://github.com/microsoft/fluentui/wiki/Perf-Testing#why-are-results-listed-in-ticks-instead-of-time-units">PR Ticks</a>
    </th>
    <th>Iterations</th>
    <th>Status</th>
  </tr><tr>
            <td>Avatar</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Avatar-mount_base.html">592</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Avatar-mount.html">588</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Button</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Button-mount_base.html">279</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Button-mount.html">290</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>Field</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Field-mount_base.html">1038</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Field-mount.html">1023</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProvider</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProvider-mount_base.html">641</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProvider-mount.html">653</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-mount_base.html">72</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-mount.html">72</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-virtual-rerender_base.html">66</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-virtual-rerender.html">68</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>FluentProviderWithTheme</td>
            <td>virtual-rerender-with-unmount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount_base.html">70</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/FluentProviderWithTheme-virtual-rerender-with-unmount.html">70</a></td>
            <td>10</td>
            <td></td>
           </tr>
<tr>
            <td>InfoButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount_base.html">9</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount.html">11</a></td>
            <td>5000</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/InfoButton-mount.regression.txt">Possible regression</a></td>
           </tr>
<tr>
            <td>MakeStyles</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/MakeStyles-mount_base.html">852</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/MakeStyles-mount.html">837</a></td>
            <td>50000</td>
            <td></td>
           </tr>
<tr>
            <td>Persona</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Persona-mount_base.html">1608</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/Persona-mount.html">1583</a></td>
            <td>5000</td>
            <td></td>
           </tr>
<tr>
            <td>SpinButton</td>
            <td>mount</td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/SpinButton-mount_base.html">1318</a></td>
            <td><a href="https://fluentuipr.z22.web.core.windows.net/pull/28335/perf-test-react-components/SpinButton-mount.html">1279</a></td>
            <td>5000</td>
            <td></td>
           </tr></table></p></details>

<div id="perfComment9425"/>

**Comment by fabricteam** [comment]
## 📊 Bundle size report

| Package & Exports                                                                                                     |      Baseline (minified/GZIP) |                            PR |                                                                                                                                                                                                  Change |
| :-------------------------------------------------------------------------------------------------------------------- | ----------------------------: | ----------------------------: | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------: |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/DataGrid.fixture.js'>DataGrid</abbr>                         | `156.306 kB`<br />`42.785 kB` | `156.496 kB`<br />`42.819 kB` | `190 B` <br />`34 B`  |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableAsDataGrid.fixture.js'>Table as DataGrid</abbr>         |  `131.96 kB`<br />`33.791 kB` |  `132.15 kB`<br />`33.825 kB` | `190 B` <br />`34 B`  |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSelectionOnly.fixture.js'>Table (Selection only)</abbr> |  `77.561 kB`<br />`19.178 kB` |  `77.751 kB`<br />`19.224 kB` | `190 B` <br />`46 B`  |

<details>
<summary>Unchanged fixtures</summary>

| Package & Exports                                                                                                                                                             |          Size (minified/GZIP) |
| ----------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | ----------------------------: |
| <samp>global-context</samp> <br /> <abbr title='bundle-size/createContext.fixture.js'>createContext</abbr>                                                                    |          `510 B`<br />`330 B` |
| <samp>global-context</samp> <br /> <abbr title='bundle-size/createContextSelector.fixture.js'>createContextSelector</abbr>                                                    |          `537 B`<br />`342 B` |
| <samp>react-accordion</samp> <br /> <abbr title='bundle-size/Accordion.fixture.js'>Accordion (including children components)</abbr>                                           |  `88.246 kB`<br />`26.712 kB` |
| <samp>react-alert</samp> <br /> <abbr title='bundle-size/Alert.fixture.js'>Alert</abbr>                                                                                       |  `82.552 kB`<br />`21.771 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/Avatar.fixture.js'>Avatar</abbr>                                                                                    |    `47.7 kB`<br />`14.504 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroup.fixture.js'>AvatarGroup</abbr>                                                                          |   `15.682 kB`<br />`6.306 kB` |
| <samp>react-avatar</samp> <br /> <abbr title='bundle-size/AvatarGroupItem.fixture.js'>AvatarGroupItem</abbr>                                                                  |  `63.876 kB`<br />`18.984 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/Badge.fixture.js'>Badge</abbr>                                                                                       |   `23.591 kB`<br />`7.264 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/CounterBadge.fixture.js'>CounterBadge</abbr>                                                                         |   `24.493 kB`<br />`7.565 kB` |
| <samp>react-badge</samp> <br /> <abbr title='bundle-size/PresenseBadge.fixture.js'>PresenceBadge</abbr>                                                                       |   `22.213 kB`<br />`7.872 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/Button.fixture.js'>Button</abbr>                                                                                    |    `36.778 kB`<br />`9.51 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/CompoundButton.fixture.js'>CompoundButton</abbr>                                                                    |   `43.932 kB`<br />`10.99 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/MenuButton.fixture.js'>MenuButton</abbr>                                                                            |  `40.965 kB`<br />`10.705 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/SplitButton.fixture.js'>SplitButton</abbr>                                                                          |   `49.197 kB`<br />`12.27 kB` |
| <samp>react-button</samp> <br /> <abbr title='bundle-size/ToggleButton.fixture.js'>ToggleButton</abbr>                                                                        |   `55.06 kB`<br />`11.446 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.All.fixture.js'>Card - All</abbr>                                                                                |   `89.11 kB`<br />`25.236 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/Card.fixture.js'>Card</abbr>                                                                                          |  `83.764 kB`<br />`23.685 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardFooter.fixture.js'>CardFooter</abbr>                                                                              |      `9.229 kB`<br />`3.9 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardHeader.fixture.js'>CardHeader</abbr>                                                                              |   `11.402 kB`<br />`4.683 kB` |
| <samp>react-card</samp> <br /> <abbr title='bundle-size/CardPreview.fixture.js'>CardPreview</abbr>                                                                            |   `10.034 kB`<br />`4.248 kB` |
| <samp>react-checkbox</samp> <br /> <abbr title='bundle-size/Checkbox.fixture.js'>Checkbox</abbr>                                                                              |  `33.164 kB`<br />`10.699 kB` |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Combobox.fixture.js'>Combobox (including child components)</abbr>                                                 |   `86.98 kB`<br />`28.084 kB` |
| <samp>react-combobox</samp> <br /> <abbr title='bundle-size/Dropdown.fixture.js'>Dropdown (including child components)</abbr>                                                 |  `85.321 kB`<br />`27.675 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ButtonProviderAndTheme.fixture.js'>react-components: Button, FluentProvider & webLightTheme</abbr>              |  `65.162 kB`<br />`17.952 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/MultipleComponents.fixture.js'>react-components: Accordion, Button, FluentProvider, Image, Menu, Popover</abbr> | `210.395 kB`<br />`58.717 kB` |
| <samp>react-components</samp> <br /> <abbr title='bundle-size/ProviderAndTheme.fixture.js'>react-components: FluentProvider & webLightTheme</abbr>                            |  `36.395 kB`<br />`11.996 kB` |
| <samp>react-datepicker-compat</samp> <br /> <abbr title='bundle-size/DatePicker.fixture.js'>DatePicker Compat</abbr>                                                          | `221.924 kB`<br />`59.118 kB` |
| <samp>react-dialog</samp> <br /> <abbr title='bundle-size/Dialog.fixture.js'>Dialog (including children components)</abbr>                                                    |  `92.138 kB`<br />`27.487 kB` |
| <samp>react-divider</samp> <br /> <abbr title='bundle-size/Divider.fixture.js'>Divider</abbr>                                                                                 |   `17.477 kB`<br />`6.357 kB` |
| <samp>react-field</samp> <br /> <abbr title='bundle-size/Field.fixture.js'>Field</abbr>                                                                                       |   `18.003 kB`<br />`6.931 kB` |
| <samp>react-image</samp> <br /> <abbr title='bundle-size/Image.fixture.js'>Image</abbr>                                                                                       |    `11.55 kB`<br />`4.627 kB` |
| <samp>react-infobutton</samp> <br /> <abbr title='bundle-size/InfoButton.fixture.js'>InfoButton</abbr>                                                                        | `131.191 kB`<br />`40.269 kB` |
| <samp>react-infobutton</samp> <br /> <abbr title='bundle-size/InfoLabel.fixture.js'>InfoLabel</abbr>                                                                          | `134.656 kB`<br />`41.338 kB` |
| <samp>react-input</samp> <br /> <abbr title='bundle-size/Input.fixture.js'>Input</abbr>                                                                                       |   `24.219 kB`<br />`7.781 kB` |
| <samp>react-label</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Label</abbr>                                                                                       |   `10.175 kB`<br />`4.243 kB` |
| <samp>react-link</samp> <br /> <abbr title='bundle-size/Label.fixture.js'>Link</abbr>                                                                                         |   `12.375 kB`<br />`5.113 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.fixture.js'>Menu (including children components)</abbr>                                                          |  `132.388 kB`<br />`40.52 kB` |
| <samp>react-menu</samp> <br /> <abbr title='bundle-size/Menu.Selectable.fixture.js'>Menu (including selectable components)</abbr>                                             | `135.152 kB`<br />`41.012 kB` |
| <samp>react-overflow</samp> <br /> <abbr title='bundle-size/hooks.fixture.js'>hooks only</abbr>                                                                               |   `12.505 kB`<br />`4.626 kB` |
| <samp>react-persona</samp> <br /> <abbr title='bundle-size/Persona.fixture.js'>Persona</abbr>                                                                                 |  `54.621 kB`<br />`16.435 kB` |
| <samp>react-popover</samp> <br /> <abbr title='bundle-size/Popover.fixture.js'>Popover</abbr>                                                                                 | `119.781 kB`<br />`36.844 kB` |
| <samp>react-portal</samp> <br /> <abbr title='bundle-size/Portal.fixture.js'>Portal</abbr>                                                                                    |    `11.82 kB`<br />`4.391 kB` |
| <samp>react-portal-compat</samp> <br /> <abbr title='bundle-size/PortalCompat.fixture.js'>PortalCompatProvider</abbr>                                                         |    `6.473 kB`<br />`2.196 kB` |
| <samp>react-positioning</samp> <br /> <abbr title='bundle-size/usePopper.fixture.js'>usePositioning</abbr>                                                                    |   `24.249 kB`<br />`8.856 kB` |
| <samp>react-progress</samp> <br /> <abbr title='bundle-size/ProgressBar.fixture.js'>ProgressBar</abbr>                                                                        |   `13.927 kB`<br />`5.488 kB` |
| <samp>react-provider</samp> <br /> <abbr title='bundle-size/FluentProvider.fixture.js'>FluentProvider</abbr>                                                                  |   `18.115 kB`<br />`6.719 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/Radio.fixture.js'>Radio</abbr>                                                                                       |   `26.952 kB`<br />`8.608 kB` |
| <samp>react-radio</samp> <br /> <abbr title='bundle-size/RadioGroup.fixture.js'>RadioGroup</abbr>                                                                             |   `11.362 kB`<br />`4.753 kB` |
| <samp>react-select</samp> <br /> <abbr title='bundle-size/Select.fixture.js'>Select</abbr>                                                                                    |   `24.915 kB`<br />`8.708 kB` |
| <samp>react-slider</samp> <br /> <abbr title='bundle-size/Slider.fixture.js'>Slider</abbr>                                                                                    |  `34.358 kB`<br />`11.107 kB` |
| <samp>react-spinbutton</samp> <br /> <abbr title='bundle-size/SpinButton.fixture.js'>SpinButton</abbr>                                                                        |  `33.452 kB`<br />`10.296 kB` |
| <samp>react-spinner</samp> <br /> <abbr title='bundle-size/Spinner.fixture.js'>Spinner</abbr>                                                                                 |   `21.363 kB`<br />`7.021 kB` |
| <samp>react-switch</samp> <br /> <abbr title='bundle-size/Switch.fixture.js'>Switch</abbr>                                                                                    |   `29.472 kB`<br />`9.226 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/PrimitivesOnly.fixture.js'>Table (Primitives only)</abbr>                                                            |  `44.652 kB`<br />`12.468 kB` |
| <samp>react-table</samp> <br /> <abbr title='bundle-size/TableSortOnly.fixture.js'>Table (Sort only)</abbr>                                                                   |  `76.891 kB`<br />`18.989 kB` |
| <samp>react-tags</samp> <br /> <abbr title='bundle-size/tag.fixture.js'>Tag</abbr>                                                                                            |   `23.153 kB`<br />`7.922 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Default.fixture.js'>Text - Default</abbr>                                                                        |   `12.563 kB`<br />`4.972 kB` |
| <samp>react-text</samp> <br /> <abbr title='bundle-size/Text.Wrappers.fixture.js'>Text - Wrappers</abbr>                                                                      |   `15.713 kB`<br />`5.293 kB` |
| <samp>react-textarea</samp> <br /> <abbr title='bundle-size/Textarea.fixture.js'>Textarea</abbr>                                                                              |    `27.69 kB`<br />`9.125 kB` |
| <samp>react-tooltip</samp> <br /> <abbr title='bundle-size/Tooltip.fixture.js'>Tooltip</abbr>                                                                                 |  `47.319 kB`<br />`16.618 kB` |
| <samp>react-utilities</samp> <br /> <abbr title='bundle-size/SSRProvider.fixture.js'>SSRProvider</abbr>                                                                       |          `180 B`<br />`159 B` |

</details>
<sub>🤖 This report was generated against <a href='https://github.com/microsoft/fluentui/commit/0d59eb984877a98a37c4aac51f885ca78ebfd849'>0d59eb984877a98a37c4aac51f885ca78ebfd849</a></sub>
<div id="bundleSizeComment9423"/>

**Comment by fabricteam** [comment]
## 🕵 fluentuiv9 No visual regressions between this PR and main
<div id="vrtCommentfluentuiv9"/>

File: packages/react-components/react-utilities/src/selection/useSelection.ts [hunk 1]
```diff
@@ -0,0 +1,33 @@
+import * as React from 'react';
+import { useControllableState } from '../hooks/useControllableState';
+import { SelectionHookParams, SelectionItemId } from './types';
+import { createSetFromIterable } from '../utils/createSetFromIterable';
+import { createSelectionManager } from './createSelectionManager';
+
+export function useSelection({
```
Comment (ling1726) [inline packages/react-components/react-utilities/src/selection/useSelection.ts:7]: can we add some tests for this?

**Review by ling1726 — APPROVED** [review]
Would be nice to have some tests for the selection utils now that they should be reusable

**Pull request merged**